### ISSUE TITLE
JSON library for modules

### ISSUE THREAD

**Issue by rossabaker** [http4s/http4s#584]
We have codec support for argonaut, circe, and json4s.  Download stats suggest all three are used in the wild.  That's fine.  They're small, and we are not an opinionated stack.

The trouble begins with some of the more specialized support modules that are emerging:
- The topic/oauth2 branch depends on circe.
- The AWS project from #583 depends on json4s-native.

This is bad for building a cohesive app.  What should we do?
1. Pick a favorite JSON library?  Not sure all our committers agree, and certainly not all our users do.
2. We already have jawn for parsing.  Add something crude for printing?
3. rapture-json has backends for all the major libraries... but would we consider that as picking a favorite?

Discuss.

**Comment by bryce-anderson** [comment]
__As to the question of the issue:__
For modules that need JSON I think we should pick one and stick with it: either the lightest or the most used. I'd personally pick the most used.

__Digression.__
I'm hoping this problem can go away once we hit something tagged 1.0. I've been a bit more vocal lately about things that are 'not in scope' and I believe even JSON should be out of scope. However, nothing gets done without JSON so as long as we don't commit to a stable API we need to keep them around so this project isn't a useless lump of bits.

After 1.0 I say they are all spun off into their own repos.

**Comment by rossabaker** [comment]
Most used: today, by a wide margin, argonaut.  Circe is newer, so it may be too early to measure.

Lightest: argonaut 6.2 has the fewest dependencies. circe depends on cats.  json4s-native depends on a deprecated stdlib feature, and json4s-jackson depends on, well, jackson.

**Comment by hvesalai** [comment]
I think argonaut is the way to go here. But ofcourse the binding shouldn't be so deep and tight that the user wouldn't be able to easily change to one of the others.

**Comment by rossabaker** [comment]
There's a difference between JSON implementation (e.g. to (de)serialize case classes in the public API) and as a member of the API itself.  In the former case, it's not a big deal.  In the latter case, things get a bit more opinionated.  It's probably best to try to avoid the latter case.

I had forgotten that rho uses json4s.  Something to consider if argonaut is our "default."

**Comment by rossabaker** [comment]
As the project moves toward Typelevel status, choosing circe might be best for interop within that ecosystem.  Still, I am most persuaded by the low-dependency argument for argonaut.  I might argue for circe later if we end up on cats instead of scalaz or shims, but argonaut seems the right choice for now.

**Comment by refried** [comment]
:+1:

**Comment by lukewyman** [comment]
Is this issue still open? If I understand correctly, the task is to find all usages of json libraries that are not argonaut and rework those parts to use argonaut, such that argonaut is the only json dependency. If that's the case, I'd like to work on it.

**Comment by rossabaker** [comment]
Great!  We'd love the help!

I think this is yet another thing that ties into #390.  The above discussion happened when we were on argonaut-6.2.x.  The community requested a reversion back to 6.1, which means that arognaut needs scalaz again.  We need that anyway today, but not necessarily in the future.

1. We can proceed right now with argonaut-6.1, and roll it forward to 6.2 when we go to fs2.  
2. We can reconsider circe, which has huge momentum, is a fellow Typelevel project, and adds no additional dependencies once core moves to cats.

I am leaning toward circe being the correct long-term answer.  Thoughts?

**Comment by bryce-anderson** [comment]
I haven't personally used Circe, but being an Argonaut fork in sure the
experience is similar (good). I agree with all the points about momentum
and ecosystem.

**Comment by lukewyman** [comment]
@rossabaker - I would  agree with you on circe. argonaut is the most widely used, but since the momentum seems to be towards cats (the main motivation behind circe), I'm inclined to think circe is the best choice. I'd like to lay claim to this task. Will fork and start looking it over today.

**Comment by rossabaker** [comment]
Wonderful, thank you!  I think at this point it hits examples and the rho project the hardest.  Things like the proposed oauth2 module are where this would expand its reach.

**Comment by rossabaker** [comment]
I think all our argonaut docs have moved to circe.  Please reopen if I'm wrong.
